### ISSUE TITLE
Add recursive template copying

### ISSUE THREAD

**Issue by dellis23** [ansible/ansible-modules-core#159]
##### Issue Type:

"Feature Idea"
##### Ansible Version:

1.7.1
##### Environment:

Running on Mac OSX and managing Ubuntu 12.04
##### Summary:

This actually goes along with a couple of other issues we've had lately (and part of what I'm mentioning is related to https://github.com/ansible/ansible/pull/8533).  Our team has a setup that seems pretty basic, but is difficult to pull off in Ansible.  We have a number of files that we need copied, some of which are encrypted.  For instance:

```
/dir/
    /private/
        private_file_1
        private_file_2
        ...
        private_file_99
```

Ideally, we'd be able to just `copy` /dir/ and be done with it, but since we need them encrypted, they won't be copied over.  Unfortunately, copying encrypted files doesn't work, so we're forced to instead create template files for all of these, store the contents in variables, and encrypt those files.  Even more unfortunately, `template` isn't recursive, so we now have to duplicate a line for each of these files to copy them!

So what's ultimately started as a simple one line copy operation has now turned into dozens of additional copy incantations, another layer of indirection in the form of variables with encrypted content, and templates that do nothing but repeat the defined variables.  It's completely destroying our attempts to keep things not repeated.
##### Steps To Reproduce:

N/A
##### Expected Results:

N/A
##### Actual Results:

N/A

**Comment by damncabbage** [comment]
I threw together something for our own setup that might be helpful: https://gist.github.com/damncabbage/12af8915e12fbf71b897

The key bit that you might be missing is that instead of creating templates, you can use `content` from the Copy action:

``` yaml
- name: "Some Files"
  copy:
    dest: "/tmp/keys/{{ item }}.key"
    content: "{{ lookup('file', 'keys/' + item + '.key.enc') | decrypt_with(vaulted_key_password) }}"
  with_items:
  - "example_one"
  - "example_two"

# ...
# Note the "copy:" hash, instead of the usual Ansible key=value pairs.
# See https://github.com/ansible/ansible/issues/9067
```

(Alternatively, have a look at [DebOps](https://github.com/debops/debops/), in particular [debops.secret](https://github.com/debops/ansible-secret).)

**Comment by daniilyar** [comment]
+1, we are also interested in 'recursive template' solution. We use Ansible for apps deployment. Deployment of several dependant apps/services to a single instance for us is a very common action. And our apps have many config files laying at different subfolders. What we need is a way to template all that app configs to remote instance recursively, as a single module call. 

We tried to implement several 'recursive template' solutions using 'with_fileglob' loops and using shell+Ansible mixes (like below). But all that implementations are very ugly and also VERY slow, espessially if there are many files to template. We tried to define as 'bulk-template' role, but it is seems to be very inconvenient to use as Ansible does not allow to call role from another role.

```
- name: Get relative paths of source directories
  local_action: shell find {{ local_files_path }}/* -type d | xargs -0 sh -c 'for filename; do prefix="{{ local_files_path }}/" && echo "${filename#$prefix}"; done' arg
  register: directory_relative_paths

- name: Get source file relative paths
  local_action: shell find "{{ local_files_path }}" -type f | xargs sh -c 'for filename; do prefix="{{ local_files_path }}/" && echo "${filename#$prefix}"; done' arg
  register: file_relative_paths

- name: Create destination directories manually as Ansible template module cannot create directories
  file: dest={{ destination_folder }}/{{ item }} state=directory
  with_items: directory_relative_paths.stdout_lines

- name: Template files recursively into remote folder
  template: src="{{ local_files_path }}/{{ item }}" dest="{{ destination_folder }}/{{ item }}"
  with_items: file_relative_paths.stdout_lines
```

After all that experiments I see that best way to implement  'recursive bulk template' feature we need is to write a module for that (or update an existing one).

I see 3 ways:
- Add 'recursive:yes' option to existing 'template' module and allow it to use directories as source;
- Write new 'bulk-template' module with behaviour similar to 'copy' module (if you put a '/' at the end of 'src' directory path, all contents of that directory will be copied to remote instance recursively);
  - Add  'template=yes' option to existing 'copy' module.

I'll try to write my own implementation (I like the 3rd way more), but I would be glad if anybody of repo admins suggest what fits better into existing Ansible code first.

**Comment by bcoca** [comment]
condensing into just 2 tasks, using lookups to find the dirs and files:

```
- file: state=directory target=/top/path/{{item}} state present
  with_items: "{{ lookup('pipe', ' find templates/ -type d')|replace("templates"/","") }}"

- template: src={{item}} target=/top/path/{{item}}
  with_items: "{{ lookup('pipe', 'find templates/ -type f')|replace("templates"/","") }}"
```

**Comment by robinrocky** [comment]
Thanks bcoca for the soln.

I have enhanced a bit more to meet specific requirement to remove .j2 at the end and retain the recursive structure in the templates folder.

Thought of adding it back here so that someone needs may use the same.

```
     - name: Copying the templated jinja2 files
       template: src={{item}} dest={{RUN_TIME}}/{{ item | regex_replace(role_path+'/templates','') | regex_replace('\.j2', '') }}
       with_items: "{{ lookup('pipe','find {{role_path}}/templates -type f').split('\n') }}"
```

**Comment by dagwieers** [comment]
You may want to look into the filetree lookup plugin I proposed in ansible/ansible#14332 and ansible/ansible#14628.

It allows for more interesting use-cases since it returns a lot of file properties that can be used for selection or manipulation. (Eg. you can use it for id-mapping when migrating NFS3/NFS4 setups)

**Comment by RussellLuo** [comment]
@bcoca Thanks for your tips!

But I guess there are some issues in your solution, and the following one works for me:

```
- name: Copy all directories recursively
  file: dest=/top/path/{{ item|replace('templates/', '') }}  state=directory       
  with_items: "{{ lookup('pipe', 'find templates/ -type d').split('\n') }}"

- name: Copy all files recursively
  template: src={{ item }} dest=/top/path/{{ item|replace('templates/', '') }}  
  with_items: "{{ lookup('pipe', 'find templates/ -type f').split('\n') }}"
```

**Comment by SeerKan** [comment]
There is a cleaner/easier way to do this with with_fileglob:

  template: src={ item }} dest=/path/on/server/{{ item|basename }} owner=user group=group mode=0664 
  with_fileglob: "path/to/templates/*"

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by jctanner** [comment]
I think we have multiple solutions for this request in the comments and now also a filetree lookup from @dagwieers.

**Comment by jeffrey4l** [comment]
When having plenty of files in one folder need template. It is slow to use the with_items solution. I still expect Ansible can have the recursive template folder solution.

**Comment by dagwieers** [comment]
@SeerKan `with_fileglob` is _not_ recursive.

@jeffrey4l  Agreed, it would be nice to have. I looked at implementing it once, but it's not as straightforward as one would think. Using with_items (or rather `with_filetree`) sufficed for us.

**Comment by adrianhowchin** [comment]
Why is this issue closed? It is a legitimate request, and would be very helpful for me too.

**Comment by romain-dartigues** [comment]
@adrianhowchin < I understand the feature has been merged through ansible/ansible#14332 and will (hopefully) be available in ansible 2.2.
I'm waiting for it too.
